pkg/kubectl/resource/visitor.go

```diff
@@ -0,0 +1,422 @@
+// of an already completed REST call.
+type Info struct {
+	Client    RESTClient
+	Mapping   *meta.RESTMapping
+	Namespace string
+	Name      string
+
+	// Optional, this is the most recent value returned by the server if available
+	runtime.Object
+	// Optional, this is the most recent resource version the server knows about for
+	// this type of resource. It may not match the resource version of the object,
+	// but if set it should be equal to or newer than the resource version of the
+	// object (however the server defines resource version).
+	ResourceVersion string
+}
+
+// NewInfo returns a new info object
+func NewInfo(client RESTClient, mapping *meta.RESTMapping, namespace, name string) *Info {
+	return &Info{
+		Client:    client,
+		Mapping:   mapping,
+		Namespace: namespace,
+		Name:      name,
+	}
+}
+
+// Visit implements Visitor
+func (i *Info) Visit(fn VisitorFunc) error {
+	return fn(i)
+}
+
+func (i *Info) Get() error {
+	obj, err := NewHelper(i.Client, i.Mapping).Get(i.Namespace, i.Name)
+	if err != nil {
+		return err
+	}
+	i.Object = obj
+	i.ResourceVersion, _ = i.Mapping.MetadataAccessor.ResourceVersion(obj)
+	return nil
+}
+
+// Watch returns server changes to this object after it was retrieved.
+func (i *Info) Watch(resourceVersion string) (watch.Interface, error) {
+	return NewHelper(i.Client, i.Mapping).WatchSingle(i.Namespace, i.Name, resourceVersion)
+}
+
+// ResourceMapping returns the mapping for this resource and implements ResourceMapping
+func (i *Info) ResourceMapping() *meta.RESTMapping {
+	return i.Mapping
+}
+
+// VisitorList implements Visit for the sub visitors it contains. The first error
+// returned from a child Visitor will terminate iteration.
+type VisitorList []Visitor
+
+// Visit implements Visitor
+func (l VisitorList) Visit(fn VisitorFunc) error {
+	for i := range l {
+		if err := l[i].Visit(fn); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// EagerVisitorList implements Visit for the sub visitors it contains. All errors
+// will be captured and returned at the end of iteration.
+type EagerVisitorList []Visitor
+
+// Visit implements Visitor, and gathers errors that occur during processing until
+// all sub visitors have been visited.
+func (l EagerVisitorList) Visit(fn VisitorFunc) error {
+	errs := []error(nil)
+	for i := range l {
+		if err := l[i].Visit(func(info *Info) error {
+			if err := fn(info); err != nil {
+				errs = append(errs, err)
+			}
+			return nil
+		}); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return util.SliceToError(errs)
+}
+
+// PathVisitor visits a given path and returns an object representing the file
+// at that path.
+type PathVisitor struct {
+	*Mapper
+	// The file path to load
+	Path string
+	// Whether to ignore files that are not recognized as API objects
+	IgnoreErrors bool
+}
+
+func (v *PathVisitor) Visit(fn VisitorFunc) error {
+	data, err := ioutil.ReadFile(v.Path)
```

Comment:
Trying to understand the approach here...

Are these really general visitors, or just collectors? My mental model is that we need to collect lots of input sources, parse out their lists of objects and merge them into a big object stream, and then provide general visitation of the objects, for iterative generation, transformation, and the final operation.